clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
         VPRINT(this, 1, "Read %zu recorded records for output #%d\n",
                outputs_[i].record.size(), i);
     }
+    // See if there was more data in the file (we do this after reading to not
+    // mis-report i/o or path errors as this error).
+    std::string err = options_.schedule_replay_istream->open_component(
+        recorded_schedule_component_name(static_cast<output_ordinal_t>(outputs_.size())));
+    if (err.empty()) {
+        error_string_ = "Not enough output streams for recorded file";
+        return STATUS_ERROR_INVALID_PARAMETER;
+    }
     for (int i = 0; i < static_cast<output_ordinal_t>(outputs_.size()); ++i) {
-        if (i < static_cast<input_ordinal_t>(inputs_.size()) &&
-            !outputs_[i].record.empty()) {
+        if (!outputs_[i].record.empty()) {
             set_cur_input(i, outputs_[i].record[0].key.input);
         } else
             set_cur_input(i, INVALID_INPUT_ORDINAL);
     }
     return STATUS_SUCCESS;
 }
 
+template <typename RecordType, typename ReaderType>
+typename scheduler_tmpl_t<RecordType, ReaderType>::scheduler_status_t
+scheduler_tmpl_t<RecordType, ReaderType>::read_traced_schedule()
+{
+    if (options_.replay_as_traced_istream == nullptr)
+        return STATUS_ERROR_INVALID_PARAMETER;
+
+    schedule_entry_t entry(0, 0, 0, 0);
+    // See comment in read_recorded_schedule() on our assumption that we can
+    // easily fit the whole context switch sequence in memory.
+    // We do not have a subfile listing feature in archive_istream_t, but we can
+    // read sequentially as each record has a cpu field.
+    // This schedule_entry_t format doesn't have the stop instruction ordinal (as it was
+    // designed for skip targets only), so we take two passes to get that information.
+    // We also need to translate the thread and cpu id values into 0-based ordinals.
+    std::unordered_map<memref_tid_t, input_ordinal_t> tid2input;
+    for (int i = 0; i < static_cast<input_ordinal_t>(inputs_.size()); ++i) {
+        tid2input[inputs_[i].tid] = i;
+    }
+    std::unordered_map<uint64_t, output_ordinal_t> cpu2output;
+    std::vector<std::map<uint64_t, uint64_t>> start2stop(inputs_.size());
+    // We number the outputs according to their order in the file.
+    // XXX: Should we support some direction from the user on this?
```

Comment:
You mean the core_id -> output_id mapping? Does that  matter? I assumed no as the ids are an internal detail. If it does, does it also matter which thread_id is mapped to which input_id?

Reply:
Simulation may want to preserve the NUMA relationships.  E.g., for a 2-socket chip, keep the simulated cores separated into their respective sockets to match how it was traced.  So the simulator will want to know which output streams are from which original cpuid.  We do have a proposal to add a cpuid query to the memtrace_stream_t interface for convenience but some of the output stream to core mapping is probably best done up front in the simulator before reading any actual stream records.

Reply:
I don't think the input mappings matter; just the output.

Reply:
I upgraded the XXX to a TODO: i#5843 and elaborated in the comment.